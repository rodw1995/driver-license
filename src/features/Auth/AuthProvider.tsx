import { useApolloClient } from '@apollo/react-hooks';
import Auth, { CognitoUser } from '@aws-amplify/auth';
import useCancelablePromise from '@rodw95/use-cancelable-promise';
import useDidMount from '@rooks/use-did-mount';
import React, { useCallback, useMemo, useRef, useState } from 'react';
import createContext from '../../utils/createContext';

export type AuthContextValue = {
  state: 'idle' | 'signingIn' | 'signedIn' | 'signingOut' | 'signedOut',
  user?: CognitoUser,
  signIn: (email: string, password: string) => Promise<CognitoUser>,
  signOut: () => Promise<void>,
}

const Context = createContext<AuthContextValue>();
export const useAuthContext = Context.hook;

type AuthProviderProps = {
  children: React.ReactNode,
}

export default ({ children }: AuthProviderProps) => {
  const apolloClient = useApolloClient();
  const makeCancelable = useCancelablePromise();

  const userRef = useRef<CognitoUser | undefined>(undefined);
  const [state, setState] = useState<AuthContextValue['state']>('idle');

  const handleUserPromise = useCallback((promise: Promise<CognitoUser>) => makeCancelable(promise)
    .then((user) => {
      userRef.current = user;
      setState('signedIn');

      return user;
    })
    .catch((err) => {
      userRef.current = undefined;
      setState('signedOut');

      throw err;
    }), [makeCancelable]);

  const signIn = useCallback<AuthContextValue['signIn']>((email, password) => {
    setState('signingIn');

    return handleUserPromise(Auth.signIn({ username: email, password }));
  }, [handleUserPromise]);

  const signOut = useCallback(() => {
    setState('signingOut');

    return makeCancelable(Auth.signOut())
      .then(() => {
        userRef.current = undefined;
        setState('signedOut');

        // Reset apollo store on logout
        apolloClient.resetStore();
      })
      .catch((err) => {
        setState('signedIn');

        throw err;
      });
  }, [makeCancelable, apolloClient]);

  // Try to fetch the authenticated user on mount
  useDidMount(() => {
    handleUserPromise(Auth.currentAuthenticatedUser()).catch(() => {});
  });

  const value = useMemo(() => ({
    state,
    user: userRef.current,
    signIn,
    signOut,
  }), [signIn, signOut, state]);

  return (
    <Context.Provider value={value}>
      {children}
    </Context.Provider>
  );
};
