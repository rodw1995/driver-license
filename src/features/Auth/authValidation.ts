import { ValidationOptions } from 'react-hook-form';

type AuthValidationRules = {
  email: ValidationOptions,
  password: ValidationOptions,
  newPassword: ValidationOptions,
  passwordConfirmation: (password: string) => ValidationOptions,
  confirmationCode: ValidationOptions,
};

const passwordValidation = {
  required: 'Password is required',
};

const authValidationRules: AuthValidationRules = {
  email: {
    required: 'Email is required',
    pattern: { value: /^\S+@\S+$/i, message: 'Not a valid email' },
  },
  password: passwordValidation,
  newPassword: {
    ...passwordValidation,
    minLength: { value: 8, message: 'Password must be at least 8 characters long' },
  },
  passwordConfirmation: (password: string) => ({
    validate: (value) => value === password || 'Password doesn\'t match',
  }),
  confirmationCode: {
    required: 'Code is required',
  },
};

export default authValidationRules;
