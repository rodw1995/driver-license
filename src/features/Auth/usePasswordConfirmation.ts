import useDidUpdate from '@rooks/use-did-update';
import { FormContextValues } from 'react-hook-form';

type PasswordConfirmationFormData = {
  password: string,
  passwordConfirmation: string,
}

export default <T extends PasswordConfirmationFormData>({
  watch, formState, errors, triggerValidation,
}: FormContextValues<T>) => {
  const password = watch('password');

  useDidUpdate(() => {
    if (formState.isSubmitted && !errors.password) {
      triggerValidation('passwordConfirmation');
    }
  }, [password]);

  return password;
};
