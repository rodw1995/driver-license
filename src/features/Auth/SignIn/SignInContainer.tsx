import React, { useCallback } from 'react';
import useCancelablePromise from '../../../hooks/useCancelablePromise';
import { useAuthContext } from '../AuthProvider';
import SignInPage, { FormData } from './SignInPage';

const SignInContainer = () => {
  const makeCancelable = useCancelablePromise();
  const { signIn } = useAuthContext();

  const onSignIn = useCallback(
    ({ email, password }: FormData) => makeCancelable(signIn(email, password)),
    [makeCancelable, signIn],
  );

  return (
    <SignInPage signIn={onSignIn} />
  );
};

export default SignInContainer;
