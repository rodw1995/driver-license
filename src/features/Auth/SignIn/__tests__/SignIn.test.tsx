import Auth from '@aws-amplify/auth';
// @ts-ignore
import { fireEvent, getByLabelText, getByRole, getByText, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { act } from 'react-test-renderer';
import AuthProvider, { AuthContextValue, useAuthContext } from '../../AuthProvider';
import SignInContainer from '../SignInContainer';

jest.mock('notistack', () => ({
  useSnackbar: () => ({
    enqueueSnackbar: () => {},
  }),
}));

jest.mock('@apollo/react-hooks', () => ({
  useApolloClient: () => ({
    resetStore: () => {},
  }),
}));

jest.mock('@aws-amplify/auth', () => ({
  __esModule: true,
  default: {
    currentAuthenticatedUser: jest.fn().mockResolvedValue(Promise.resolve(null)),
    signIn: jest.fn(),
    signOut: () => Promise.resolve(),
  },
}));

describe('SignIn', () => {
  // Keep track of the auth context value to test
  let authContextValue: AuthContextValue;
  const AuthContextSetter = () => {
    authContextValue = useAuthContext();
    return null;
  };

  // Setup providers and SignInContainer
  beforeEach(async () => {
    jest.resetAllMocks();

    act(() => {
      render(
        <AuthProvider>
          <AuthContextSetter />
          <MemoryRouter>
            <SignInContainer />
          </MemoryRouter>
        </AuthProvider>,
      );
    });

    await waitFor(() => expect(Auth.currentAuthenticatedUser).toHaveBeenCalledTimes(1));

    expect(authContextValue.state).toBe('signedOut');
    expect(authContextValue.user).toBe(undefined);
  });

  it('displays a sign in form', () => {
    const form = screen.getByTestId('sign-in-form');

    expect(form).toBeTruthy();
    expect(form.innerHTML).toMatchSnapshot();
    expect(getByLabelText(form, 'Email Address')).toBeTruthy();
    expect(getByLabelText(form, 'Password')).toBeTruthy();
    expect(getByRole(form, 'button')).toBeTruthy();
    expect(getByRole(form, 'button').textContent).toBe('Sign In');
  });

  it('shows required fields errors', async () => {
    const form = screen.getByTestId('sign-in-form');

    fireEvent.click(getByRole(form, 'button'));

    await waitFor(() => Promise.all([
      expect(getByText(form, 'Email is required')).toBeTruthy(),
      expect(getByText(form, 'Password is required')).toBeTruthy(),
    ]));

    expect(authContextValue.state).toBe('signedOut');
    expect(authContextValue.user).toBe(undefined);
  });

  it('shows an unknown email address error', async () => {
    const form = screen.getByTestId('sign-in-form');

    fireEvent.change(getByLabelText(form, 'Email Address'), {
      target: { value: 'robbie@example.com' },
    });

    fireEvent.change(getByLabelText(form, 'Password'), {
      target: { value: 'password' },
    });

    fireEvent.click(getByRole(form, 'button'));

    // @ts-ignore
    // eslint-disable-next-line prefer-promise-reject-errors
    Auth.signIn.mockResolvedValue(Promise.reject({ code: 'UserNotFoundException' }));

    await waitFor(() => expect(getByText(form, 'No user with this email exists')).toBeTruthy());

    expect(authContextValue.state).toBe('signedOut');
    expect(authContextValue.user).toBe(undefined);
  });

  it('shows an incorrect password error', async () => {
    const form = screen.getByTestId('sign-in-form');

    fireEvent.change(getByLabelText(form, 'Email Address'), {
      target: { value: 'robbie@example.com' },
    });

    fireEvent.change(getByLabelText(form, 'Password'), {
      target: { value: 'password' },
    });

    fireEvent.click(getByRole(form, 'button'));

    // @ts-ignore
    // eslint-disable-next-line prefer-promise-reject-errors
    Auth.signIn.mockResolvedValue(Promise.reject({ code: 'NotAuthorizedException' }));

    await waitFor(() => expect(getByText(form, 'Incorrect password')).toBeTruthy());

    expect(authContextValue.state).toBe('signedOut');
    expect(authContextValue.user).toBe(undefined);
  });

  it('can sign in successfully', async () => {
    const form = screen.getByTestId('sign-in-form');

    fireEvent.change(getByLabelText(form, 'Email Address'), {
      target: { value: 'robbie@example.com' },
    });

    fireEvent.change(getByLabelText(form, 'Password'), {
      target: { value: 'password' },
    });

    fireEvent.click(getByRole(form, 'button'));

    const resolvedUser = {};
    // @ts-ignore
    Auth.signIn.mockResolvedValue(Promise.resolve(resolvedUser));

    await waitFor(() => expect(Auth.signIn).toHaveBeenCalledTimes(1));

    expect(authContextValue.state).toBe('signedIn');
    expect(authContextValue.user).toBe(resolvedUser);
  });
});
