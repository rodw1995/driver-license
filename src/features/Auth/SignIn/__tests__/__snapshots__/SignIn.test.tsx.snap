// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn displays a sign in form 1`] = `"<div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused\\" data-shrink=\\"true\\" for=\\"email-address\\" id=\\"email-address-label\\">Email Address</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth Mui-focused Mui-focused MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" id=\\"email-address\\" name=\\"email\\" type=\\"text\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-128 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-130 PrivateNotchedOutline-legendNotched-131\\"><span>Email Address</span></legend></fieldset></div></div><div class=\\"MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth\\"><label class=\\"MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined\\" data-shrink=\\"false\\" for=\\"password\\" id=\\"password-label\\">Password</label><div class=\\"MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl\\"><input aria-invalid=\\"false\\" autocomplete=\\"current-password\\" id=\\"password\\" name=\\"password\\" type=\\"password\\" class=\\"MuiInputBase-input MuiOutlinedInput-input\\" value=\\"\\"><fieldset aria-hidden=\\"true\\" class=\\"PrivateNotchedOutline-root-128 MuiOutlinedInput-notchedOutline\\"><legend class=\\"PrivateNotchedOutline-legendLabelled-130\\"><span>Password</span></legend></fieldset></div></div><button class=\\"MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-2 MuiButton-containedPrimary MuiButton-fullWidth\\" tabindex=\\"0\\" type=\\"submit\\"><span class=\\"MuiButton-label\\">Sign In</span><span class=\\"MuiTouchRipple-root\\"></span></button><div class=\\"MuiGrid-root MuiGrid-container\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary\\" href=\\"/forgotPassword\\">Forgot password?</a></div><div class=\\"MuiGrid-root MuiGrid-item\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-body2 MuiTypography-colorPrimary\\" href=\\"/signUp\\">No account? Sign Up</a></div></div>"`;
